graphistry/umap_utils.py

```diff
@@ -344,10 +352,13 @@ def transform_umap(self, df: pd.DataFrame,
 
     def _bundle_embedding(self, emb, index):
         # Converts Embedding into dataframe and takes care if emb.dim > 2
-        if emb.shape[1] == 2 and 'cudf.core.dataframe' not in str(getmodule(emb)):
+        if emb.shape[1] == 2 and 'cudf.core.dataframe' not in str(getmodule(emb)) and not hasattr(emb, 'device'):
             emb = pd.DataFrame(emb, columns=[config.X, config.Y], index=index)
         elif emb.shape[1] == 2 and 'cudf.core.dataframe' in str(getmodule(emb)):
             emb.rename(columns={0: config.X, 1: config.Y}, inplace=True)
+        elif emb.shape[1] == 2 and hasattr(emb, 'device'):
+            import cudf
+            emb = cudf.DataFrame(emb, columns=[config.X, config.Y], index=index)
```

Comment:
all this sniffing is suspicious, maybe we can pass in the engine?

Reply:
Daniel made those changes, I am not sure why